packages/teams-js/src/internal/videoUtils.ts

```diff
@@ -78,36 +108,142 @@ function createProcessedStreamGenerator(
+      ({ discardAudioInferenceResult }: { discardAudioInferenceResult: boolean }) => {
+        this.shouldDiscardAudioInferenceResult = discardAudioInferenceResult;
+      },
+    );
+  }
+
+  public transform = async (originalFrame, controller): Promise<void> => {
+    const timestamp = originalFrame.timestamp;
+    if (timestamp !== null) {
+      try {
+        const { videoFrame, metadata: { audioInferenceResult } = {} } = await this.extractVideoFrameAndMetadata(
+          originalFrame,
+        );
+        const frameProcessedByApp = await this.videoFrameHandler({ videoFrame, audioInferenceResult });
+        // the current typescript version(4.6.4) dosn't support webcodecs API fully, we have to do type conversion here.
+        const processedFrame = new VideoFrame(frameProcessedByApp as unknown as CanvasImageSource, {
+          // we need the timestamp to be unchanged from the oirginal frame, so we explicitly set it here.
+          timestamp: timestamp,
+        });
+        controller.enqueue(processedFrame);
+        videoFrame.close();
+        originalFrame.close();
+        (frameProcessedByApp as VideoFrame).close();
+      } catch (error) {
+        originalFrame.close();
+        this.notifyError(error);
+      }
+    } else {
+      this.notifyError('timestamp of the original video frame is null');
+    }
+  };
+
+  /**
+   * @hidden
+   * Extract video frame and metadata from the given texture.
+   *
+   * @internal
+   * Limited to Microsoft-internal use
+   */
+  private extractVideoFrameAndMetadata = async (
+    texture: VideoFrame,
+  ): Promise<{ videoFrame: VideoFrame; metadata: { audioInferenceResult?: Uint8Array } }> => {
+    if (inServerSideRenderingEnvironment()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    if (texture.format !== 'NV12') {
+      this.notifyError('Unsupported video frame format');
+      throw new Error('Unsupported video frame format');
+    }
+
+    /**
+     * stream id for audio inference data
+     */
+    const AUDIO_INFERENCE_RESULT_STREAM_ID = 0x31646961;
+
+    // The rectangle of pixels to copy from the texture
+    const headerRect = { x: 0, y: 0, width: texture.codedWidth, height: 2 };
+    // it's in NV12 format, but the real data is in the y plane only.
+    const headerBuffer = new ArrayBuffer((headerRect.width * headerRect.height * 3) / 2);
+    await texture.copyTo(headerBuffer, { rect: headerRect });
+    const headerDataView = new Uint32Array(headerBuffer);
+    // const [ oneTextureId, version, frameRowOffset, frameFormat, frameWidth, frameHeight, multiStreamHeaderRowOffset, multiStreamCount ] = headerDataView;
+    const metadataRect = {
+      x: 0,
+      y: headerDataView[6], // multiStreamHeaderRowOffset
+      width: texture.codedWidth,
+      height: texture.codedHeight - headerDataView[6], // multiStreamHeaderRowOffset
+    };
+    const metadataBuffer = new ArrayBuffer((metadataRect.width * metadataRect.height * 3) / 2);
```

Comment:
Why are you multiplying by 3 and dividing by 2? There are a lot of [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) in this code, please either remove them, or store them as consts and clearly document why they need to be these specific values.

Reply:
Added comments to explain the size of NV12 format

Reply:
I appreciate the comments, but you still have magic numbers that are not consts, and there is no explanation of why you are using 2 or 3. Someone else looking at this code later would not be able to look at this code and understand what is doing.

Reply:
Still have magic numbers here. The comment above doesn't really help me. You have a line of math that shows two things being equal ((metadataRect.width * metadataRect.height) + (metadataRect.width * metadataRect.height) / 2 and (metadataRect.width * metadataRect.height * 3) / 2) but those two things aren't equal?

Reply:
I added a link to NV12 format document at line 345. The calculation is straightforward when looking at the memory layout of the NV12 format.
> // it's in NV12 format (https://learn.microsoft.com/en-us/windows/win32/medfound/recommended-8-bit-yuv-formats-for-video-rendering#nv12), so the
    // number bytes of the header is (the size of the Y plane + the size of the UV plane)

Reply:
I appreciate that to you this math is trivial. Comments and clearly named constants outlining why those values are being used are still necessary. "Go read this very long piece of external documentation" isn't sufficient. You could link to a specific part of that page which outlines how the math works, or you could explain it here. Asking other developers who have to understand this code to go to that page and search for "3" is not maintainable.

Reply:
Added the explanation in the comment